src/backend/storage/ipc/procarray.c

```diff
@@ -4929,38 +4929,188 @@ GetSessionIdByPid(int pid)
  * Set the destination group slot or group id in PGPROC, and send a signal to the proc.
  * slot is NULL on QE.
  */
-void
-ResGroupSignalMoveQuery(int sessionId, void *slot, Oid groupId)
+bool
+ResGroupMoveSignalTarget(int sessionId, void *slot, Oid groupId, bool executor)
 {
 	pid_t pid;
 	BackendId backendId;
 	ProcArrayStruct *arrayP = procArray;
+	bool sent = false;
+	bool found = false;
+
+	Assert(Gp_role == GP_ROLE_DISPATCH || Gp_role == GP_ROLE_EXECUTE);
+	Assert(groupId != InvalidOid);
+
+	if (Gp_role == GP_ROLE_EXECUTE)
+		executor = true;
 
 	LWLockAcquire(ProcArrayLock, LW_SHARED);
 	for (int i = 0; i < arrayP->numProcs; i++)
 	{
 		volatile PGPROC *proc = &allProcs[arrayP->pgprocnos[i]];
 		if (proc->mppSessionId != sessionId)
 			continue;
+		/* 
+		 * Before, we didn't distinguish entrydb processes from target.
+		 * There is a case with enrtydb executors when we can send a signal to
+		 * target process only, but not to entrydb executor process or
+		 * vice versa.
+		 * As a mediocre solution we assume executors (including entrydb) not
+		 * using queryCommandId.
+		 */
+		if ((proc->queryCommandId <= -1) ^ executor)
+			continue;
 
+		found = true;
 		pid = proc->pid;
 		backendId = proc->backendId;
-		if (Gp_role == GP_ROLE_DISPATCH)
+
+		SpinLockAcquire(&proc->movetoMutex);
+		if (Gp_role == GP_ROLE_DISPATCH && proc->queryCommandId > -1)
```

Comment:
Maybe `if (!executor)` can also work here?

Reply:
Here and in functions near or in another files we rely on `proc->queryCommandId`, but not on a local `executor` argument, so it's cleaner to keep it's usage if possible.

Reply:
Well... This function seemed to me to be overcomplicated: passed func argument 'executor' + local variable 'executor' +  proc->queryCommandId comparison
But it is probably the only way to do it safely, ~who am I~ I think colleagues know better

Reply:
Yes, it's overcomplicated. But I don't see any good option to move some logic to another (new) one.

Reply:
Additional comment added. The logic still the same.